### ISSUE TITLE
non-smooth transition and unbalanced brightness

### ISSUE THREAD

**Issue by assaft** [jcelaya/hdrmerge#65]
I'm trying to merge two DNG files and I see two problems in the merged file:
1. The transition between the exposures is not smooth. 
2. There's some exceptionally bright area in the middle.

The two exposures were taken 2 seconds apart and they are 1 EV apart. Maybe there was a slight change in the light level, but I can't tell. It's clear from the merged file that the darker file comes out (i.e. blended) too dark in the merged file (compared to the brighter file). And It's also unclear why there is an exceptionally bright region in the middle of the merged file. RawDigger says that the darker exposure has no clipping. 

Darker exposure:
![p7293435](https://cloud.githubusercontent.com/assets/4903680/8993986/338a937a-3711-11e5-9175-ed74304bb05a.jpg)

Merged file:
![p7293435-6-output](https://cloud.githubusercontent.com/assets/4903680/8994000/4a87a9aa-3711-11e5-9b2c-55e72f0c326a.jpg)

The files and the output with comments embedded can be found here:
https://dl.dropboxusercontent.com/u/21642925/MergeProblem.zip

**Comment by Beep6581** [comment]
Probably a duplicate of issue 6.

**Comment by heckflosse** [comment]
Merges fine with this fix: https://github.com/heckflosse/hdrmerge/commit/49cbeae8678ba170de14bbb6c35700dcb89a2799

**Comment by assaft** [comment]
Hi @heckflosse, I haven't had a chance to use your fix but I checked how things look with align/crop unchecked in v0.5 (the latest release). The result is much better, which suggests that the problem is largely due to over-correction of the aligner. There is much last clipping and the transition between the bright area and the darker areas in the sky is smooth. See here:
![tests](https://cloud.githubusercontent.com/assets/4903680/10005915/a545c432-60c2-11e5-88fd-a67039b2bf44.png)
I take it that your fix solves also the issue with the remaining clipped highlights. Great!
It is actually my fault for not disabling the alignment&crop options. I used a tripod and I knew my bracketed shots are perfectly aligned. But I think that in other cases people may take the same picture handheld and then they will need HDRMerge's aligner to help (even just a bit). For example, to match the horizon line. So if the aligner is confused by the waves or something else, it won't be usable just like we see here.

**Comment by heckflosse** [comment]
Hi @assaft ,
here's how it looks like now. Merged without alignment and crop using my hdrmerge fork. Resulting dng opened in rawtherapee with neutral profile. Still some visible banding...

http://filebin.net/gn9esy6q6f/example.png

Ingo

**Comment by assaft** [comment]
Hi Ingo,
Thanks again!
The highlights look better, but as you say - there is some visible banding in the sky. This is something I don't see with version 0.5 when align&crop are unchecked (see the photo I posted above).
Could you please post a link to the merged file (to make it easier to compare the fork vs 0.5)?
Assaf

**Comment by heckflosse** [comment]
Sure, uploading, takes some minutes...

**Comment by heckflosse** [comment]
@assaft : http://filebin.net/694830q503

**Comment by heckflosse** [comment]
@assaft btw, my fork also contains this commit from Javier https://github.com/jcelaya/hdrmerge/commit/7502607346684a747cb0ff8e2b00edff85b89f57 which is not in version 5.0

**Comment by heckflosse** [comment]
@assaft For reference I also uploaded the output of hdrmerge branch master https://github.com/jcelaya/hdrmerge/commits/master to above mentioned filebin link. So you can see what's fixed in my fork...

Ingo

**Comment by assaft** [comment]
@heckflosse it looks like we get very different results from version 0.5. This is what I got when I merged with no align, no crop, 16bps, full size preview, blur radius 3: http://filebin.net/tilmpwisrr; I'm running HDRMerge on Windows 7 64bit.

**Comment by heckflosse** [comment]
@assaft https://github.com/jcelaya/hdrmerge/commits/master is not 0.5!!! It's one commit ahead of 0.5 and this commit is the main reason for the differences.
Your result (from 0.5) looks really good and maybe Javier should review his latest commit. I only fixed the clearly wrong behaviour of this commit...

**Comment by assaft** [comment]
@heckflosse Oh OK. Sorry I confused 0.5 and the master branch. But one more thing - do you see the difference between 0.5 and the fork version? I see a problem with non-smooth transition in the fork version. If this is what you meant when you mentioned 'banding' before than I guess that we talk about the same thing. But to be on the safe side I prepared a comparison. It requires to zoom-in to see the transition that I'm talking about.
![tests2](https://cloud.githubusercontent.com/assets/4903680/10008132/1133e2a8-60d4-11e5-8eca-f240871802e1.png)

**Comment by heckflosse** [comment]
@assaft : Yes, that's exactly what I meant with 'banding'. It's not caused by my bug fix, but by Javier's change to the handling of dark images with his latest commit. Maybe I also find a solution for this...

**Comment by assaft** [comment]
@heckflosse OK, good, so we are talking about the same thing. So there are the problems that I see (assuming that align and crop are unchecked):
v0.5: clipped highlights in the middle.
master: bug in the response function of dark images + clipped highlights in the middle + banding.
fork:  clipped highlights in the middle + banding.

**Comment by heckflosse** [comment]
@assaft agreed, but I assume that Javier had a reason to change the response function for dark images, so we have to take this also into account (Currently don't know which Issue)

**Comment by Beep6581** [comment]
The reason could be https://github.com/jcelaya/hdrmerge/issues/64 or https://github.com/jcelaya/hdrmerge/issues/52#issuecomment-87461099

**Comment by jcelaya** [comment]
About the implementation of the response function:
One of the ideas behind HDRMerge is that merging in RAW space is easier and faster because we can assume a linear response function from the camera... but, that's not always true. Some cameras have a small non-linearity in the high end of the response function. So, I decided to use a spline for the 25% lighter values, and a linear function for the darker ones.

The problem with dark images is that there might not be enough information to build the spline. In that case, the fallback is to use a linear response function for the full range. That's what that last commit added.

In any case, response functions are always built from incomplete data (because an image may not have pixels of every possible brightness, and there is noise...), and they are an approximation of the true camera response function. So, improvements are always possible (and welcome).

**Comment by assaft** [comment]
Javier, do you consider the banding [shown here](https://github.com/jcelaya/hdrmerge/issues/65#issuecomment-142139308) an expected result of the last commit or a bug?
Also, what do you think about  the clipped highlights in the middle of the bright area?

**Comment by fanckush** [comment]
Is this problem still present? if so then can someone show me a recent sample with the problem?
